src/nebari/hookspecs.py

```diff
@@ -1,54 +1,37 @@
-import contextlib
-import pathlib
-from typing import Any, Dict, List
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import ClassVar, Optional
 
-import pydantic
-import typer
 from pluggy import HookimplMarker, HookspecMarker
+from pydantic import BaseModel
+from typer import Typer
 
 from nebari import schema
 
 hookspec = HookspecMarker("nebari")
 hookimpl = HookimplMarker("nebari")
 
 
-class NebariStage:
-    name: str = None
-    priority: int = None
+class NebariStage(ABC):
+    name: ClassVar[str]
+    priority: ClassVar[int]
+    input_schema: ClassVar[Optional[BaseModel]] = None
+    output_schema: ClassVar[BaseModel]
 
-    input_schema: pydantic.BaseModel = None
-    output_schema: pydantic.BaseModel = None
-
-    def __init__(self, output_directory: pathlib.Path, config: schema.Main):
+    def __init__(self, output_directory: Path, config: schema.Main):
         self.output_directory = output_directory
         self.config = config
 
-    def render(self) -> Dict[str, str]:
-        return {}
```

Comment:
The return type should be `Dict[Union[str, Path], str]`. Then we need to use the overload decorator in the derived class.

Reply:
Is this something you'll handle in this PR?

Reply:
There is no elegant way to do this currently. I think it's better to left those abstract method without return time.

Reply:
The return type should actually be `Dict[Path, str]`.